src/main/java/org/opensearch/knn/index/memory/NativeMemoryLoadStrategy.java

```diff
@@ -119,6 +123,55 @@ public NativeMemoryAllocation.IndexAllocation load(NativeMemoryEntryContext.Inde
             );
         }
 
+        @Override
+        public NativeMemoryAllocation.IndexAllocation load(NativeMemoryEntryContext.IndexEntryContext indexEntryContext)
+            throws IOException {
+            final Path absoluteIndexPath = Paths.get(indexEntryContext.getKey());
+            final KNNEngine knnEngine = KNNEngine.getEngineNameFromPath(absoluteIndexPath.toString());
+            if (knnEngine != KNNEngine.FAISS) {
+                // We will support other non-FAISS native engines (ex: NMSLIB) soon.
+                return loadWithAbsoluteIndexPath(indexEntryContext);
+            }
+
+            final FileWatcher fileWatcher = new FileWatcher(absoluteIndexPath);
+            fileWatcher.addListener(indexFileOnDeleteListener);
+            fileWatcher.init();
+
+            final Directory directory = indexEntryContext.getDirectory();
+
+            // Ex: Input -> /a/b/c/_0_NativeEngines990KnnVectorsFormat_0.vec
+            // Output -> _0_NativeEngines990KnnVectorsFormat_0.vec
+            final String logicalIndexPath = absoluteIndexPath.getFileName().toString();
+
+            final int indexSize = (int) directory.fileLength(logicalIndexPath);
+
+            try (IndexInput readStream = directory.openInput(logicalIndexPath, IOContext.READONCE)) {
+                IndexInputWithBuffer indexInputWithBuffer = new IndexInputWithBuffer(readStream);
+                long indexAddress = JNIService.loadIndex(indexInputWithBuffer, indexEntryContext.getParameters(), knnEngine);
+
+                SharedIndexState sharedIndexState = null;
+                String modelId = indexEntryContext.getModelId();
+                if (IndexUtil.isSharedIndexStateRequired(knnEngine, modelId, indexAddress)) {
+                    log.info("Index with model: \"{}\" requires shared state. Retrieving shared state.", modelId);
+                    sharedIndexState = SharedIndexStateManager.getInstance().get(indexAddress, modelId, knnEngine);
+                    JNIService.setSharedIndexState(indexAddress, sharedIndexState.getSharedIndexStateAddress(), knnEngine);
+                }
+
+                final WatcherHandle<FileWatcher> watcherHandle = resourceWatcherService.add(fileWatcher);
+                return new NativeMemoryAllocation.IndexAllocation(
+                    executor,
+                    indexAddress,
+                    indexSize / 1024,  // Convert bytes in KB unit.
+                    knnEngine,
+                    absoluteIndexPath.toString(),
+                    indexEntryContext.getOpenSearchIndexName(),
+                    watcherHandle,
+                    sharedIndexState,
+                    IndexUtil.isBinaryIndex(knnEngine, indexEntryContext.getParameters())
+                );
```

Comment:
1. You can move this code outside of try with resource block.
2. This logic is similar to what we have in `loadWithAbsoluteIndexPath` so lets move them to a private function.

Reply:
yeah, good idea!
Will do in the next rev.